Warnings cleanup (#1385)
mono/linker: docs/error-codes.md

- The assembly 'assembly' in `PreserveDependency` attribute could not be resolved.

  ```C#
  // IL2003: Could not resolve dependency assembly 'NonExistentAssembly' specified in a 'PreserveDependency' attribute
  [PreserveDependency("MyMethod", "MyType", "NonExistentAssembly")]
  void TestMethod()
  {
  }
  ```

#### `IL2004`: Could not resolve dependency type 'type' specified in a 'PreserveDependency' attribute

- The type 'type' in `PreserveDependency` attribute could not be resolved.

  ```C#
  // IL2004: Could not resolve dependency type 'NonExistentType' specified in a 'PreserveDependency' attribute
  [PreserveDependency("MyMethod", "NonExistentType", "MyAssembly")]
  void TestMethod()
  {
  }
  ```

#### `IL2005`: Could not resolve dependency member 'member' declared in type 'type' specified in a 'PreserveDependency' attribute

- The member 'member' in `PreserveDependency` attribute could not be resolved.

  ```C#
  // IL2005: Could not resolve dependency member 'NonExistentMethod' declared on type 'MyType' specified in a 'PreserveDependency' attribute
  [PreserveDependency("NonExistentMethod", "MyType", "MyAssembly")]
  void TestMethod()
  {
  }
  ```

> #### `IL2006` Trim analysis: The requirements declared via the `DynamicallyAccessedMembersAttribute` on <value description> don't match those on <target description>. The source value must declare at least the same requirements as those declared on the target location it's assigned to

Reviewer: What is the benefit of `Trim analysis` in the warning message for the developers?

Author: This is the subcategory - I'm adding it to this doc both as documentation and as a way to quickly review the subcategories. It will directly map to the `TrimAnalysis` property in MSBuild (which will suppress these messages).

Reviewer: hmm, not sure this is best option. Maybe we could add Category bellow to not break the search

Author: I don't understand the concern? Is it about the ability to search this document?

Note, the output of the linker looks like this:
```
File.cs(11,22): Trim analysis warning IL2006: The requirements ....
```

Reviewer: Yep, I think if I copy `IL2006: The requirements` text it should find me the doc about the warning. I don't see a value in having `Trim analysis` in the warning output. It just makes the output verbose and harder to read as different warnings have different formats and the information about Trim analysis is or could be already included in the warning code

- The property 'property' has DynamicallyAccessedMembersAttribute on it, but the linker could not determine the backing fields for the property to propagate the attribute to the field.

> #### `IL2043`: Trying to propagate DynamicallyAccessedMemberAttribute from property 'property' to its setter 'method', but it already has such attribute.

Author: yes - I guess we could check if they're different and warn about that, but in general there needs to be a warning in this place.

  {
      // IL2056 Trim analysis: The assembly 'NonExistentAssembly' can not be found
      Activator.CreateInstance("NonExistentAssembly", "MyType");
  }
  ```

#### `IL2057` Trim analysis: The type 'type name' can not be found in assembly 'assembly'

- Calling `CreateInstance` with a type name which can't be resolved in the target assembly.  

  ``` C#
  void TestMethod()
  {
      // IL2057 Trim analysis: The type 'NonExistentType' can not be found in assembly 'MyAssembly'
      Activator.CreateInstance("MyAssembly", "NonExistentType");
  }
  ```

#### `IL2058` Trim analysis: Unrecognized value passed to the parameter 'typeName' of method 'CreateInstance'. It's not possible to guarantee the availability of the target type.

- The value passed as the type name to the `CreateInstance` method can't be statically analyzed, ILLink can't make sure that the type is available.  

  ``` C#
  void TestMethod(string typeName)
  {
      // IL2058 Trim analysis: Unrecognized value passed to the parameter 'typeName' of method 'CreateInstance'. It's not possible to guarantee the availability of the target type.
      Activator.CreateInstance("MyAssembly", typeName);
  }
  ```

> #### `IL2059` Trim analysis: Unrecognized value passed to the parameter 'assemblyName' of method 'CreateInstance'. It's not possible to guarantee the availability of the target type.

Reviewer 2: Just checking - we want to share 2056-2059 across various CreateInstance* methods right?

Author: They're already shared between AppDomain/Assembly/Activator CreateInstance methods

#### `IL2049`: Unrecognized internal attribute 'attribute'

- The internal attribute name 'attribute' being used in the xml is not supported by the linker, check the spelling and the supported internal attributes.

> #### `IL2050` Trim analysis: Making a generic type instantiation from 'type' which has 'DynamicallyAccessedMembersAttribute' on some of its generic parameters. ILLink currently doesn't analyze type values for generic parameters when making a generic type instantiation via 'System.Type.MakeGenericType'

Reviewer: Could you unify quotes around DynamicallyAccessedMembersAttribute and other names. It looks like there are for some reasons used with "`", "'" and with no quotes

Reviewer: It looks like the warning message is missing the reason.

```suggestion
#### `IL2050` Making a generic type instantiation from 'type' which has 'DynamicallyAccessedMembersAttribute' on some of its generic parameters is not supported.
```

Author: I tried to unify it like this:
* In messages it should be single-quoted : 'DynamicallyAccessedAttribute' - just like any other reference to names from the code (all method names, type names and so on should also be single-quoted)
* In the documentation it should be backtick-quoted so that the docs show it formatted as code.
* There should be no double quotes anywhere

Author: As for the message - I don't think we should say "not supported" - it is supported in the code, it's just that linker doesn't analyze it.

Reviewer: ok but I think the sentence needs tweaking as "Making a generic type instantiation from 'type' which has 'DynamicallyAccessedMembersAttribute' on some of its generic parameters" is weird

- Analysis of type parameters for `System.Type.MakeGenericType` is not yet implemented. If the open generic type `type` has `DynamicallyAccessedMembersAttribute` on any of its generic parameters, ILLink currently can't validate that the requirements are fulfilled by the calling method.  

  ``` C#
  class Lazy<[DynamicallyAccessedMembers(DynamicallyAccessedMemberType.PublicParameterlessConstructor)] T> 
  {
      // ...
  }
  
  void TestMethod()
  {
      // IL2050 Trim analysis: Making a generic type instantiation from 'Lazy<T>' which has 
      // 'DynamicallyAccessedMembersAttribute' on some of its generic parameters. ILLink currently doesn't analyze 
      // type values for generic parameters when making a generic type instantiation via 'System.Type.MakeGenericType'
      typeof(Lazy<>).MakeGenericType(new Type[] { typeof(TestType) });
  }
  ```

>   #### `IL2051` Trim analysis: Unrecognized type value used to called the method 'MakeGenericType'. It's not possible to guarantee the availability of requirements of such type.

Reviewer: I don't understand the warning message :-(

Author: As per the code sample - calling `MakeGenericType` on a `Type` instance we can't statically analyze - if that type has annotations on its generic parameters, we would be required to validate those - but can't since we don't know the type.
If anybody has a better idea about the message, I'm more than happy to change the text.

Reviewer: How is that different from IL2052? Do we need to expose the limitation over warning codes?

Does it make sense to tweak IL2052 to cover all such cases?

```
#### `IL2052` Method `MakeGenericType(...)` is called on instance which cannot be determined.
```

Author: That message is also not correct - but what about this:
```
IL2051 Trim analysis: Call to `System.Reflection.MethodInfo.MakeGenericType` can not be statically analyzed. It's not possible to guarantee the availability of requirements of the generic type.
```

Basically the same as IL2056 but for types. And to avoid future question - I do want to keep the `MakeGenericType` and `MakeGenericMethod` warnings separate. If nothing else, to give us freedom of improving the warning message if the current one is not good enough (and sharing it between both cases would make it much harder to come up with a message)

Reviewer: Yep, this is better

- If `System.Type.MakeGenericType` is called on an type which is not statically analyzable. If the actual type has generic parameters with `DynamicallyAccessedMembersAttribute` ILLink would be required to fulfill the requirements declared by those attributes, but since the ILLink doesn't know the type, it can't determine if such requirements exist.  

  ``` C#
  void TestMethod(Type type)
  {

>       // IL2051 Trim analysis: Unrecognized value passed to the parameter 'typeArguments' of method 'System.Type.MakeGenericType'

Reviewer: Should we copy the message here (it can go out of sync) ?

Author: I was trying to show how the message would look with specific references to code - since the message with parametrized pieces can be hard to understand.

      type.MakeGenericType(new Type[] { typeof(TestType) });
  }
  ```

> #### `IL2052` Trim analysis: Unrecognized value passed to the parameter 'typeName' of method 'System.Type.GetType(Type typeName)'. It's not possible to guarantee the availability of the target type.

Reviewer: ```suggestion
#### `IL2052` Argument type used for parameter 'typeName' of method 'System.Type.GetType(Type typeName)' cannot be determined.`
```

Author: I find the new message confusing - "Argument type" to me means the declared type of the argument, which in this case will pretty always be `System.Type` - and that one was determined (otherwise linker would not have a problem with it). It's the value passed to the argument which we could not figure out.
I guess there's also a question of terminology - argument or parameter?

Reviewer: I think this is quite clear. From Wikipedia

> Unlike argument in usual mathematical usage, the argument in computer science is thus the actual input expression passed/supplied to a function, procedure, or routine in the invocation/call statement, whereas the parameter is the variable inside the implementation of the subroutine. 

I agree we need to settle on how we are going to identify the type and we are not consistent in other messages. In this case, it could be `Argument instance used for parameter ....`

Author: Thanks for the explanation. I will take a note and do this sweeping change in a separate PR (this is already getting out of hand). I will have another PR focusing just on the messages anyway.

- If the type name passed to the `System.Type.GetType` is statically known ILLink can make sure it's preserved and the application code will work after trimming. But if the type name is unknown, it could point to a type which ILLink will not see being used anywhere else and would remove it from the application, potentially breaking the application.  

  ``` C#
  void TestMethod()
  {
      string typeName = ReadName();

      // IL2052 Trim analysis: Unrecognized value passed to the parameter 'typeName' of method 'System.Type.GetType(Type typeName)'
      Type.GetType(typeName);
  }
  ```

> #### `IL2053` Trim analysis: 'Assembly.CreateInstance' is not supported with trimming. Use 'Type.GetType' and `Activator.CreateInstance` instead

Reviewer: ```suggestion
#### `IL2053` Arguments used for 'System.Assembly.CreateInstance(...)' method cannot be analyzed. Consider using `System.Activator.CreateInstance` method instead.
```

- ILLink currently doesn't analyze assembly instances and thus it doesn't know on which assembly the `Assembly.CreateInstance` was called. ILLink has support for `Type.GetType` instead, for cases where the parameter is a string literal. The result of which can be passed to `Activator.CreateInstance` to create an instance of the type.  

  ``` C#
  void TestMethod()
  {
      // IL2053 Trim analysis: 'Assembly.CreateInstance' is not supported with trimming. Use 'Type.GetType' instead
      AssemblyLoadContext.Default.Assemblies.First(a => a.Name == "MyAssembly").CreateInstance("MyType");

      // This can be replaced by
      Activator.CreateInstance(Type.GetType("MyType, MyAssembly"));
  }
  ```
  

> #### `IL2054` Trim analysis: Unrecognized value passed to the parameter 'type' of method 'System.Runtime.CompilerServices.RuntimeHelpers.RunClassConstructor'. It's not possible to guarantee the availability of the target static constructor.

Reviewer: We should be consistent in the warning message. If we include method name we should include it in the same form in all warnings (e.g. with parameters and parameter names)

Author: I tried to not use full formatting the message titles as these are all parametrized in code. The message in the sample tries to be better about the formatting but still not perfect (I didn't run linker on the code to get the exact string it would produce).

      // It's not possible to guarantee the availability of the target static constructor.
      RuntimeHelpers.RunClassConstructor(type.TypeHandle);
  }
  ```

#### `IL2055` Trim analysis: Call to `System.Reflection.MethodInfo.MakeGenericMethod` can not be statically analyzed. It's not possible to guarantee the availability of requirements of the generic method.

- ILLink currently doesn't analyze `MethodInfo` values and thus can't statically determine the generic method the `MakeGenericMethod` operates on. If the actual method has generic parameters with `DynamicallyAccessedMembersAttribute` ILLink would be required to fulfill the requirements declared by those attributes, but since the ILLink doesn't know the method, it can't determine if such requirements exist.  

  ``` C#
  void TestMethod()
  {
      // IL2055 Trim analysis: Call to `System.Reflection.MethodInfo.MakeGenericMethod` can not be statically analyzed. It's not possible to guarantee the availability of requirements of the generic method.
      typeof(MyType).GetMethod("MyMethod").MakeGenericMethod(new Type[] { typeof(MyType) });
  }
  ```

> #### `IL2056` Trim analysis: The assembly 'assembly name' can not be found

Reviewer: ```suggestion
#### `IL2056` The assembly name 'name' used in method call 'method' references assembly which is not available.
```

- Calling `CreateInstance` with assembly name 'assembly name' which can't be resolved.  

  ``` C#
  void TestMethod()
  {
      // IL2056 Trim analysis: The assembly 'NonExistentAssembly' can not be found
      Activator.CreateInstance("NonExistentAssembly", "MyType");
  }
  ```

#### `IL2058` Trim analysis: Unrecognized value passed to the parameter 'typeName' of method 'CreateInstance'. It's not possible to guarantee the availability of the target type.

- The value passed as the type name to the `CreateInstance` method can't be statically analyzed, ILLink can't make sure that the type is available.  

  ``` C#
  void TestMethod(string typeName)
  {
      // IL2058 Trim analysis: Unrecognized value passed to the parameter 'typeName' of method 'CreateInstance'. It's not possible to guarantee the availability of the target type.
      Activator.CreateInstance("MyAssembly", typeName);
  }
  ```

> #### `IL2059` Trim analysis: Unrecognized value passed to the parameter 'assemblyName' of method 'CreateInstance'. It's not possible to guarantee the availability of the target type.

Reviewer: Why not use IL2058 for this?

Author: Good point - fixed.

- The value passed as the assembly name or path to the `CreateInstance` method can't be statically analyzed, ILLink can't make sure that the type is available.  

  ``` C#
  void TestMethod(string assemblyName)
  {
      // IL2059 Trim analysis: Unrecognized value passed to the parameter 'assemblyName' of method 'CreateInstance'. It's not possible to guarantee the availability of the target type.
      Activator.CreateInstance(assemblyName, "MyType");
  }
  ```

> #### `IL2060` Trim analysis: Unrecognized type value passed to <target description>. It's not possible to guarantee that the requirements declared by the `DynamicallyAccessedMembersAttribute` are met.

Reviewer: I don't understand this warning message.:-(

Author: As per the example, it is used when a value which we can't statically analyze is assigned to something (target) which has data flow annotation - and thus linker can't guarantee that the annotation's requirement is fulfilled. I welcome any suggestions to improve the message.

Reviewer: I think this could be something like

```
#### `IL2059` Argument instance used for parameter 'x' in method 'm' doesn't need to meet DynamicallyAccessedMembersAttribute requirements.
```

Author: What about:
```
Value passed to <target description> can not be statically determined and may not meet 'DynamicallyAccessedMembersAttribute' requirements.
```

So specifically for a method parameter:
```
Value passed to the parameter 'x' of method 'm' can not be statically determined and may not meet 'DynamicallyAccessedMembersAttribute' requirements.
```

Reviewer: I think that's better but `Value passed to the parameter 'x' of method 'm'` should have its own warning code because if someone will be searching for `Value passed to the parameter` it probably won't find this reference easily

Author: There are currently 5 cases:
* `parameter 'name' of method 'name'`
* `return value of method 'name'`
* `field 'name'`
* `implicit 'this' parameter of method 'name'`
* `generic parameter 'name' from 'type/method'`

You think we should have separate code for each of them?

Reviewer: I'd say yes as they are about construction and would allow us to have better warning message for each case

Author: What about 2047 - it uses the same pattern. For example:
```
IL2047: 'DynamicallyAccessedMemberTypes' in 'DynamicallyAccessedMembersAttribute' on parameter 'type' on method 'Derived.TestMethod' don't match overridden parameter 'type' on method 'Base.TestMethod'. All overridden members must have the same 'DynamicallyAccessedMembersAttribute' usage.
```

Reviewer: Yeah, that will need the same treatment
